source/common/network/connection_impl.cc

```diff
@@ -66,7 +66,10 @@ ConnectionImpl::ConnectionImpl(Event::Dispatcher& dispatcher, ConnectionSocketPt
       fd(), [this](uint32_t events) -> void { onFileEvent(events); }, Event::FileTriggerType::Edge,
       Event::FileReadyType::Read | Event::FileReadyType::Write);
 
-  transport_socket_->setTransportSocketCallbacks(*this);
+  // Callers who pass 'connected' as 'false' must do this call themselves, e.g., when fully constructed.
+  if (connected) {
```

Comment:
This is pretty fragile IMO. WDYT about making the constructor private and then have a static factory function that does the right thing for the various places that need to construct a connection? Would that be possible?

Reply:
That would be a bit more "boilerplate" but would be not fragile at all, I'll give it a go. And sorry about the format thing, getting rusty...